### PR TITLE
Fix deterministic MVID and add PdbChecksum

### PR DESCRIPTION
**Pull request by vitek-karas**
Make deterministic MVID unique even when the only change is in PE headers. Implement deterministic PDB ID generation independent of MVID and implement writing of PdbCheckum debug header entry.

Observable differences (other than the above):
* ISymbolWriter now writes the PDB in `GetDebugHeader` (well except for MPDB writer which I didn't change as it's really tricky). Callers should not notice much of a difference if they assumed only `Dispose` would actually write.
* `Module.Write(stream` now requires the stream to be read/write. It already did so for cases where strong name was written, now it requires it also when deterministic MVID is turned on.

Deterministic MVID
So far the deterministic MVID was computed as a hash of the IL metadata. It didn't include PE headers.
This means that if there's an assembly which only differs in the PE architecture it would have the same MVID.
This can happen relatively easily for facades (assemblies with only type forwarders) as those are typically
identical across platforms, possibly except for the architecture.
The correct way to calculate MVID is to hash the content of the entire file with the MVID itself and
the strong name signature (if presenet) zeroed out.
This change modified the MVID computation to work that way: Write the entire image with MVID all zeroes,
compute the hash, write the MVID and then compute the strong name.

PdbChecksum
In order for the produced PDB to be fully verifiable the assembly must contain a PdbChecksum
debug header entry which has a cryptographic hash of the associated PDB. The wayt to calculate the checksum
is prescribed as other programs must be able to validate it. See https://github.com/dotnet/runtime/blob/main/docs/design/specs/PE-COFF.md#pdb-checksum-debug-directory-entry-type-19.
This change implements that behavior. Symbols are now written explicitly once all metadata is processed
but before the final assembly image is written. For portable PDBs the full file is written with PDB ID
set to all zeroes. Then the crypto hash is calculated (using SHA256).
The hash is then written into the PdbChecksum debug header entry of the assembly image.

Deterministic Portable PDB ID
So far the PDB ID was set to the same value as MVID. With the fix for MVID described above, this doesn't
work anymore. In case of an embedded PDB, it's not possible to calculate MVID using the hash
with the PDB embedded, as the PDB would now contain the MVID (cycle).
The recommended way to calculate PDB ID is to use the same mechanism as for calculating PDB checksum
and use the first 20 bytes of the has as the PDB ID.
This change implements this by using the first 20 bytes of the hash and sets the PDB ID as the last
change of the portable PDB.
The calculated PDB ID is also then set into the CodeView debug header entry of the assembly.

Added tests for reading and writing both portable and embedded portable PDBs with checksums.
Added test to validate stability of PDB ID across multiple writes.
Added test to validate stability of MVID across multiple writes and the fact that it changes
when just the PE architecture is changed.

This is a fix for https://github.com/dotnet/linker/issues/2203.

### CODE REVIEW

File: Mono.Cecil.Cil/PortablePdb.cs [hunk 1]
```diff
@@ -171,7 +171,7 @@ public ISymbolReader GetSymbolReader (ModuleDefinition module, string fileName)
 				throw new InvalidOperationException ();
 
 			return new EmbeddedPortablePdbReader (
-				(PortablePdbReader) new PortablePdbReaderProvider ().GetSymbolReader (module, GetPortablePdbStream (entry)));
+				(PortablePdbReader)new PortablePdbReaderProvider ().GetSymbolReader (module, GetPortablePdbStream (entry)));
```
Comment (marek-safar) [inline Mono.Cecil.Cil/PortablePdb.cs:174]: Please undo any line damage or formating changes, they make upstream merges pain

**Review by marek-safar — CHANGES_REQUESTED** [review]

File: Mono.Cecil.Cil/PortablePdb.cs (hunk 1 above)
Comment (vitek-karas) [reply Mono.Cecil.Cil/PortablePdb.cs]: Apparently I hit something in VS which reformatted the entire file. I'm so glad we have lint in dotnet/linker :wink:.
Anyway - hopefully I was able to revert all of these now.

**Comment by vitek-karas** [comment]
This should be fully ready now.

**Review by marek-safar — APPROVED** [review]
I'd suggest to make upstream PR after you merge it here

**Comment by vitek-karas** [comment]
Ported this to jbevain/cecil as https://github.com/jbevain/cecil/pull/810.

**Pull request merged**